internal/controller/controllers/agent_controller.go

```diff
@@ -80,69 +97,244 @@ func (r *AgentReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
 	if err = r.Get(ctx, kubeKey, clusterDeployment); err != nil {
 		errMsg := fmt.Sprintf("failed to get clusterDeployment with name %s in namespace %s",
 			agent.Spec.ClusterDeploymentName.Name, agent.Spec.ClusterDeploymentName.Namespace)
-		Requeue = false
-		clientError := true
-		if !k8serrors.IsNotFound(err) {
-			Requeue = true
-			clientError = false
-		}
-		clusterDeploymentRefErr := newKubeAPIError(errors.Wrapf(err, errMsg), clientError)
 		// Update that we failed to retrieve the clusterDeployment
-		r.updateFailure(ctx, agent, clusterDeploymentRefErr)
-		return ctrl.Result{Requeue: Requeue}, nil
+		return r.updateStatus(ctx, agent, nil, errors.Wrapf(err, errMsg), !k8serrors.IsNotFound(err))
 	}
 
 	// Retrieve cluster for ClusterDeploymentName from the database
 	cluster, err := r.Installer.GetClusterByKubeKey(kubeKey)
 	if err != nil {
 		if gorm.IsRecordNotFoundError(err) {
-			Requeue = true
 			inventoryErr = common.NewApiError(http.StatusNotFound, err)
+			internalError = false
 		} else {
-			Requeue = false
 			inventoryErr = common.NewApiError(http.StatusInternalServerError, err)
+			internalError = true
 		}
 		// Update that we failed to retrieve the cluster from the database
-		r.updateFailure(ctx, agent, inventoryErr)
-		return ctrl.Result{Requeue: Requeue}, nil
+		return r.updateStatus(ctx, agent, nil, inventoryErr, internalError)
+	}
+
+	//Retrieve host from cluster
+	host := getHostFromCluster(cluster, agent.Name)
+	if host == nil {
+		return r.updateStatus(ctx, agent, nil, errors.New("host not found in cluster"), false)
 	}
 
-	var result ctrl.Result
 	// check for updates from user, compare spec and update if needed
-	result, err = r.updateIfNeeded(ctx, agent, cluster)
+	err = r.updateIfNeeded(ctx, agent, cluster)
 	if err != nil {
-		r.updateFailure(ctx, agent, err)
-		return result, nil
+		return r.updateStatus(ctx, agent, host, err, !IsHTTP4XXError(err))
 	}
 
-	err = r.updateInventory(cluster, agent)
+	err = r.updateInventory(host, agent)
 	if err != nil {
-		r.updateFailure(ctx, agent, err)
-		return ctrl.Result{Requeue: true}, nil
+		return r.updateStatus(ctx, agent, host, err, true)
+	}
+
+	return r.updateStatus(ctx, agent, host, nil, false)
+}
+
+func (r *AgentReconciler) updateStatus(ctx context.Context, agent *adiiov1alpha1.Agent, h *models.Host, syncErr error, internal bool) (ctrl.Result, error) {
```

Comment:
What is the meaning or significance of `internal`? A doc string would help.

Reply:
Done